src/vs/workbench/contrib/terminalContrib/chat/browser/terminal.chatHint.contribution.ts

```diff
@@ -0,0 +1,263 @@
+	}
+
+	private _xterm: IXtermTerminal & { raw: RawXtermTerminal } | undefined;
+	private _chatHint: IDecoration | undefined;
+
+	constructor(
+		private readonly _instance: ITerminalInstance | IDetachedTerminalInstance,
+		processManager: ITerminalProcessManager | ITerminalProcessInfo,
+		widgetManager: TerminalWidgetManager,
+		@IInlineChatService private readonly _inlineChatService: IInlineChatService,
+		@IInstantiationService private readonly _instantiationService: IInstantiationService,
+		@ITerminalService private readonly _terminalService: ITerminalService
+	) {
+		super();
+	}
+
+	xtermOpen(xterm: IXtermTerminal & { raw: RawXtermTerminal }): void {
+		if (this._terminalService.instances.length !== 1) {
+			// only show for the first terminal
+			return;
+		}
+		this._xterm = xterm;
+		const capability = this._instance.capabilities.get(TerminalCapability.CommandDetection);
+		if (capability) {
+			this._register(Event.once(capability.promptInputModel.onDidStartInput)(() => this._addChatHint()));
+		} else {
+			this._register(this._instance.capabilities.onDidAddCapability(e => {
+				if (e.capability.type === TerminalCapability.CommandDetection) {
+					const capability = this._instance.capabilities.get(TerminalCapability.CommandDetection);
+					this._register(Event.once(capability!.promptInputModel.onDidStartInput)(() => this._addChatHint()));
+					if (!capability!.promptInputModel.value) {
+						this._addChatHint();
+					}
+				}
+			}));
+		}
+
+		this._register(Event.once(this._inlineChatService.onDidChangeProviders)(() => this._addChatHint()));
+	}
+
+	private _addChatHint(): void {
+		const instance = this._instance instanceof TerminalInstance ? this._instance : undefined;
+		if (!instance || !this._xterm || this._hintWidget || instance?.capabilities.get(TerminalCapability.CommandDetection)?.hasInput) {
+			return;
+		}
+
+		if (!this._chatHint) {
+			const marker = this._xterm.raw.registerMarker();
+			if (!marker) {
+				return;
+			}
+
+			if (this._xterm.raw.buffer.active.cursorX === 0) {
+				return;
+			}
+			this._register(marker);
+			this._chatHint = this._xterm.raw.registerDecoration({
+				marker,
+				x: this._xterm.raw.buffer.active.cursorX + 1
+			});
+		}
+
+		this._register(this._xterm.raw.onKey(() => this._chatHint?.dispose()));
+		this._chatHint?.onRender((e) => {
+			if (!this._hintWidget && this._xterm?.isFocused && this._terminalService.instances.length === 1) {
+				const chatProviders = [...this._inlineChatService.getAllProvider()];
+				if (chatProviders?.length) {
+					const widget = this._instantiationService.createInstance(TerminalChatHintWidget, instance);
+					this._hintWidget = widget.getDomNode(chatProviders);
+					if (!this._hintWidget) {
+						return;
+					}
+					e.appendChild(this._hintWidget);
+					e.classList.add('terminal-chat-hint');
+					e.style.width = (this._xterm!.raw.cols - this._xterm!.raw.buffer.active.cursorX) / this._xterm!.raw.cols * 100 + '%';
```

Comment:
We could avoid this math/custom CSS by passing a width in cells into `registerDecoration`?

Reply:
that handles the initial width, but if a user resizes, we'll want to update it here, right?

Reply:
(outside of the if statement)

Reply:
actually, I don't think passing in a width in cells to `registerDecoration` would work

Reply:
isn't this essentially the width of the cells to the right of the prompt? seems to work based on my testing, maybe i'm missing something though.